sql/lease.go

```diff
@@ -0,0 +1,551 @@
+		leaseKey := makeLeaseKey(s.nodeID, lease.ID, lease.Version, lease.expiration)
+		if log.V(3) {
+			log.Infof("acquire: %s", prettyKey(leaseKey, 0))
+		}
+		b := &client.Batch{}
+		b.Put(leaseKey, nil)
+		// TODO(pmattis): Setting the system DB trigger is currently necessary
+		// because the lease table resides in the system DB span. Perhaps it
+		// shouldn't.
+		txn.SetSystemDBTrigger()
+		return txn.CommitInBatch(b)
+	})
+	return lease, err
+}
+
+// Release a previously acquired table descriptor lease.
+func (s LeaseStore) Release(lease *LeaseState) error {
+	leaseKey := makeLeaseKey(s.nodeID, lease.ID, lease.Version, lease.expiration)
+	if log.V(3) {
+		log.Infof("release: %s", prettyKey(leaseKey, 0))
+	}
+	return s.db.Txn(func(txn *client.Txn) error {
+		b := &client.Batch{}
+		b.Del(leaseKey)
+		txn.SetSystemDBTrigger()
+		return txn.CommitInBatch(b)
+	})
+}
+
+// Publish a new version of a table descriptor. The update closure may be
+// called multiple times if retries occur: make sure it does not have side
+// effects.
+func (s LeaseStore) Publish(tableID ID, update func(*TableDescriptor) error) error {
+	desc := &TableDescriptor{}
+	descKey := MakeDescMetadataKey(tableID)
+
+	retryOpts := retry.Options{
+		InitialBackoff: 20 * time.Millisecond,
+		MaxBackoff:     2 * time.Second,
+		Multiplier:     2,
+	}
+
+	for r := retry.Start(retryOpts); r.Next(); {
+		// Get the current version of the table descriptor non-transactionally.
+		//
+		// TODO(pmattis): Do an inconsistent read here?
+		if err := s.db.GetProto(descKey, desc); err != nil {
+			return err
+		}
+		// Check to see if there are any leases that still exist on the previous
+		// version of the descriptor.
+		now := s.clock.Now()
+		count, err := s.countLeases(desc.ID, desc.Version-1, now.WallTime)
+		if err != nil {
+			return err
+		}
+		if count != 0 {
+			log.Infof("publish (count leases): descID=%d version=%d count=%d", desc.ID, desc.Version-1, count)
+			continue
+		}
+
+		// At this point, desc.Version is the only version of the descriptor that
+		// has leases outstanding. Lease acquisition (see acquire()) maintains the
+		// invariant that no new leases for desc.Version-1 will be granted once
+		// desc.Version exists.
+		expectedVersion := desc.Version
+		err = s.db.Txn(func(txn *client.Txn) error {
+			// Re-read the current version of the table descriptor, this time
+			// transactionally.
+			if err := txn.GetProto(descKey, desc); err != nil {
+				return err
+			}
+			if expectedVersion != desc.Version {
+				// The version changed out from under us. Someone else must be
+				// performing a schema change operation.
+				if log.V(3) {
+					log.Infof("publish (version changed): %d != %d", expectedVersion, desc.Version)
+				}
+				return errLeaseVersionChanged
+			}
+
+			// Update the descriptor.
+			if err := update(desc); err != nil {
+				return err
+			}
+
+			// Bump the version and modification time.
+			desc.Version = desc.Version + 1
```

Comment:
aren't you breaking your own back here if the retryable gets called multiple times?